pennylane_lightning/core/src/simulators/lightning_kokkos/catalyst/tests/Test_LightningKokkosSimulator.cpp

```diff
@@ -0,0 +1,662 @@
+// Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the License);
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an AS IS BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include <algorithm>
+#include <cmath>
+#include <complex>
+#include <cstddef>
+#include <limits>
+#include <type_traits>
+#include <utility>
+#include <variant>
+#include <vector>
+
+#include <catch2/catch.hpp>
+
+#include "LightningKokkosSimulator.hpp"
+#include "QuantumDevice.hpp"
+#include "TestHelpers.hpp"
+
+/// @cond DEV
+namespace {
+using namespace Catalyst::Runtime::Simulator;
+using namespace Pennylane::Util;
+using LKSimulator = LightningKokkosSimulator;
+using QDevice = Catalyst::Runtime::QuantumDevice;
+
+GENERATE_DEVICE_FACTORY(LightningKokkosSimulator,
+                        Catalyst::Runtime::Simulator::LightningKokkosSimulator);
+} // namespace
+/// @endcond
+
+/**
+ * @brief Tests the LightningKokkosSimulator class.
+ *
+ */
+TEST_CASE("LightningKokkosSimulator::constructor", "[constructibility]") {
+    SECTION("LightningKokkosSimulator") {
+        REQUIRE(std::is_constructible<LKSimulator>::value);
+    }
+    SECTION("LightningKokkosSimulator(string))") {
+        REQUIRE(std::is_constructible<LKSimulator, std::string>::value);
+    }
+}
+
+TEST_CASE("Test the device factory method", "[constructibility]") {
+    std::unique_ptr<QDevice> LKsim(LightningKokkosSimulatorFactory(""));
+    REQUIRE(LKsim->GetNumQubits() == 0);
+}
+
+TEST_CASE("LightningKokkosSimulator::unit_tests", "[unit tests]") {
+    SECTION("Managing Qubits") {
+        std::unique_ptr<LKSimulator> LKsim = std::make_unique<LKSimulator>();
+        std::vector<intptr_t> Qs = LKsim->AllocateQubits(0);
+        REQUIRE(LKsim->GetNumQubits() == 0);
+        LKsim->AllocateQubits(2);
+        REQUIRE(LKsim->GetNumQubits() == 2);
+        LKsim->AllocateQubits(2);
+        REQUIRE(LKsim->GetNumQubits() == 4);
+        LKsim->ReleaseQubit(0);
+        REQUIRE(
+            LKsim->GetNumQubits() ==
+            4); // releasing only one qubit does not change the total number.
```

Comment:
I don't know. I noticed that when adding the tests and added a comment for a follow-up discussion. @maliasadi ?

Reply:
Good question, to be honest the dynamic allocation feature is not fully baked, because there was no interest in making it usable with PennyLane. So only the `AllocateQubits` and `ReleaseAllQubits` functions are actually used.